### PR TITLE
Fix Half X86_F16 CUDA build failure

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141035
* __->__ #140994
* #140720
* #140567
* #140566
* #140565
* #140564

It passed PyTorch CI, but internally we saw failures from this.

Differential Revision: [D66137897](https://our.internmc.facebook.com/intern/diff/D66137897/)

### CODE REVIEW

File: c10/util/Half.h [hunk 1]
```diff
@@ -345,13 +345,9 @@ inline uint16_t fp16_ieee_from_fp32_value(float f) {
   return static_cast<uint16_t>(
       (sign >> 16) |
       (shl1_w > UINT32_C(0xFF000000) ? UINT16_C(0x7E00) : nonsign));
-#endif // X86_F16
+#endif // C10_X86_F16
 }
 
-#ifdef X86_F16
-#undef X86_F16
-#endif // X86_F16
```
Comment (malfet) [inline c10/util/Half.h:353]: Why don't you want to undefine it at the end of the file?

**Review by malfet — APPROVED** [review]
LGTM, sounds like you are addressing feedback left at https://github.com/pytorch/pytorch/pull/140720#discussion_r1847107150

File: c10/util/Half.h (hunk 1 above)
Comment (swolchok) [reply c10/util/Half.h:353]: I thought we were allowed to define macros as long as they were properly namespaced. Seemed unlikely to collide

Comment (Skylion007) [reply c10/util/Half.h:353]: Macros ignore C++ namespaces...

Comment (swolchok) [reply c10/util/Half.h:353]: > Macros ignore C++ namespaces...

"C10_" is the namespace in my book.

Comment (swolchok) [reply c10/util/Half.h:353]: anyway I updated the diff to undefine the macro

**Comment by malfet** [comment]
Oops, this have landed already, but it would be nice to highlight, that it not only broke internal builds, but other ecosystem libraries, fbgemm for example, see https://github.com/pytorch/FBGEMM/pull/3394#issuecomment-2486370560

**Pull request closed without merging**